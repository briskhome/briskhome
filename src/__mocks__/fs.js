'use strict';

const fs = jest.genMockFromModule('fs');

let mockFiles = Object.create(null);
function __setMockFiles(newMockFiles) {
  mockFiles = Object.create(null);
  for (const file in newMockFiles) {
    const dir = file.substring(0, file.lastIndexOf('/'));

    if (!mockFiles[dir]) {
      mockFiles[dir] = [];
    }
    mockFiles[dir].push(file.split('/').pop());
  }
}

function readdir(directoryPath, callback) {
  return callback(null, mockFiles[directoryPath] || []);
}

function realpath(...args) {
  const callback = args.pop();
  const directory = `./${args
    .join('/')
    .split('./')
    .join('')}`;
  if (directory in mockFiles) return callback(null, directory);
  return callback(new Error());
}

fs.__setMockFiles = __setMockFiles;
fs.readdir = readdir;
fs.realpath = realpath;
fs.realpath.native = realpath;

module.exports = fs;
