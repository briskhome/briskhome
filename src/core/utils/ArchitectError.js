/* @flow */

class ArchitectError extends Error {
  constructor(message: string, code: string, data: Object) {
    super(message);
    this.message = message;
    this.code = code;
    this.data = data;

    Error.captureStackTrace(this, this.constructor);
  }

  toString() {
    return this.message;
  }
}

export default ArchitectError;
