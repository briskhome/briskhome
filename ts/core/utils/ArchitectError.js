class ArchitectError extends Error {
  code: string;
  message: string;
  data: object | undefined | null;

  constructor(message: string, code: string, data: object | undefined | null = null) {
    super(message);
    this.message = message;
    this.code = code;
    this.data = data;
    Error.captureStackTrace(this, this.constructor);
  }

  toString() {
    return this.message;
  }

}

export default ArchitectError;