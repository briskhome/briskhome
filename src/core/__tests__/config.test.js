import fs from 'fs';
import Config from '../lib/config';
import ArchitectError from '../utils/ArchitectError';

jest.mock('fs');
jest.mock('path');

const LIB = {
  './lib/core.example': '',
  './lib/core.example/index.js': 'core.example',
  './lib/core.example/package.json': 'core.example',
};

const NODE_MODULES = {
  './node_modules/briskhome-example': '',
  './node_modules/briskhome-example/index.js': '',
  './node_modules/briskhome-example/package.json': '',
};

describe('core -> config', () => {
  let config;

  beforeEach(() => {
    fs.__setMockFiles({ ...LIB, ...NODE_MODULES });
    config = new Config();
  });

  it('constructor()', () => {
    const directories = ['test'];
    const config = new Config(directories);
    expect(config.directories).toEqual(directories);
  });

  it('async discover()', async () => {
    const result = await config.discover();
    expect(result).toHaveLength(2);
    expect(result[0]).toEqual(Object.keys(LIB)[0]);
    expect(result[1]).toEqual(Object.keys(NODE_MODULES)[0]);
  });

  it('async prepareExtensions()');
  xdescribe('async getMetadata()', () => {
    it('fails', async () => {
      const res = await config.getMetadata();
    });
  });
  describe('async getRealpath()', () => {
    it('resolves extension (lib)', async () => {
      const base = './lib';
      const name = './core.example';
      const result = await config.getRealpath(base, name);
      expect(result).toBeTruthy();
    });
    it('resolves extension (node_modules)', async () => {
      const base = '.';
      const name = 'briskhome-example';
      const result = await config.getRealpath(base, name);
      expect(result).toBeTruthy();
    });
    it('fails to resolve extension (lib)', async () => {
      const base = './lib';
      const name = './core.invalid';
      try {
        await config.getRealpath(base, name);
      } catch (e) {
        expect(e).toBeInstanceOf(ArchitectError);
      }
    });
    it('fails to resolve extension (node_modules)', async () => {
      const base = '.';
      const name = 'briskhome-invalid';
      try {
        await config.getRealpath(base, name);
      } catch (e) {
        expect(e).toBeInstanceOf(ArchitectError);
      }
    });
  });
  describe('parseExtensions()', () => {
    const packageJson = {
      name: 'package-name',
      dir: '/tmp/briskhome',
      main: '/tmp/briskhome/index.js',
    };

    it('single extension (short notation)', () => {
      const result = config.parseExtensions(
        { 'briskhome:extension': ['default'] },
        packageJson,
      );
      expect(result).toMatchSnapshot();
    });

    it('single extension (verbose notation)', () => {
      const result = config.parseExtensions(
        {
          'briskhome:extension': [
            {
              name: 'default',
              main: 'default.js',
              dependencies: ['dependency-name'],
            },
          ],
        },
        packageJson,
      );
      expect(result).toMatchSnapshot();
    });

    it('single extension (verbose notation / no overrides)', () => {
      const result = config.parseExtensions(
        {
          'briskhome:extension': [{}],
        },
        packageJson,
      );
      expect(result).toMatchSnapshot();
    });

    it('multiple extensions (same type)', () => {
      const result = config.parseExtensions(
        { 'briskhome:extension': ['one', 'two'] },
        packageJson,
      );
      expect(result).toMatchSnapshot();
    });

    it('multiple extensions (different types)', () => {
      const result = config.parseExtensions(
        { 'briskhome:extension-one': ['one'] },
        { 'briskhome:extension-two': ['two'] },
        packageJson,
      );
      expect(result).toMatchSnapshot();
    });

    it('invalid extension', () => {
      const result = config.parseExtensions(
        { 'briskhome:extension': [0] },
        packageJson,
      );
      expect(result).toMatchSnapshot();
    });
  });
});
