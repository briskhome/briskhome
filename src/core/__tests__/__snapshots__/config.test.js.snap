// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core -> config parseExtensions() invalid extension 1`] = `Array []`;

exports[`core -> config parseExtensions() multiple extensions (different types) 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isDefault": true,
    "main": undefined,
    "metadata": Object {
      "briskhome:extension-two": Array [
        "two",
      ],
    },
    "name": "one",
    "type": "briskhome:extension-one",
  },
]
`;

exports[`core -> config parseExtensions() multiple extensions (same type) 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isDefault": false,
    "main": "/tmp/briskhome/index.js",
    "metadata": Object {
      "dir": "/tmp/briskhome",
      "main": "/tmp/briskhome/index.js",
      "name": "package-name",
    },
    "name": "one",
    "type": "briskhome:extension",
  },
  Object {
    "dependencies": Array [],
    "isDefault": false,
    "main": "/tmp/briskhome/index.js",
    "metadata": Object {
      "dir": "/tmp/briskhome",
      "main": "/tmp/briskhome/index.js",
      "name": "package-name",
    },
    "name": "two",
    "type": "briskhome:extension",
  },
]
`;

exports[`core -> config parseExtensions() single extension (short notation) 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isDefault": true,
    "main": "/tmp/briskhome/index.js",
    "metadata": Object {
      "dir": "/tmp/briskhome",
      "main": "/tmp/briskhome/index.js",
      "name": "package-name",
    },
    "name": "default",
    "type": "briskhome:extension",
  },
]
`;

exports[`core -> config parseExtensions() single extension (verbose notation / no overrides) 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isDefault": true,
    "main": "/tmp/briskhome/index.js",
    "metadata": Object {
      "dir": "/tmp/briskhome",
      "main": "/tmp/briskhome/index.js",
      "name": "package-name",
    },
    "name": "package-name",
    "type": "briskhome:extension",
  },
]
`;

exports[`core -> config parseExtensions() single extension (verbose notation) 1`] = `
Array [
  Object {
    "dependencies": Array [
      "dependency-name",
    ],
    "isDefault": true,
    "main": "/tmp/briskhome/default.js",
    "metadata": Object {
      "dir": "/tmp/briskhome",
      "main": "/tmp/briskhome/index.js",
      "name": "package-name",
    },
    "name": "default",
    "type": "briskhome:extension",
  },
]
`;
